packages/core/test/financial-templates/common/financial-product-libraries/contract-for-difference-libraries/RangeBondContractForDifferenceFinancialProductLibrary.js

```diff
@@ -0,0 +1,156 @@
+        from: cfdMock.address,
+      });
+      assert.equal(expiraryTokensForCollateral.toString(), toWei("0.2"));
+    });
+    it("Midway between bounds should return long worth bond notional (long side is long yield dollar)", async () => {
+      // If the price is between the low and high price ranges then the payout is simply that of a yield dollar. i.e every
+      // long token is worth the bond notional of 100. At a price of 20 we are between the bounds. Each long should be worth
+      // 100 so there should be 100/20=5 UMA per long token. As each collateralPerpair is worth 10, expiryPercentLong should
+      // be 10/5=0.5, thereby allocating half to the long and half to the short.
+      const expiraryTokensForCollateral1 = await rangeBondCFDFPL.computeExpiraryTokensForCollateral.call(toWei("20"), {
+        from: cfdMock.address,
+      });
+      assert.equal(expiraryTokensForCollateral1.toString(), toWei("0.5"));
+
+      // Equally, at a price of 40 each long should still be worth 100 so there should be 100/40=2.5 UMA per long. As
+      // each collateralPerpair=10 expiryPercentLong should be 10/2.5=0.25, thereby allocating 25% to long and the remaining to short.
+      const expiraryTokensForCollateral2 = await rangeBondCFDFPL.computeExpiraryTokensForCollateral.call(toWei("20"), {
+        from: cfdMock.address,
+      });
+      assert.equal(expiraryTokensForCollateral2.toString(), toWei("0.5"));
+    });
+
+    it("Arbitary price should return correctly", async () => {
+      // Value of long token: T = min(N/P,N/R1) + max((N/R2*(P-R2))/P),0)
+      // expected payout is T/collateralPerPair. We can compute the exact numerical amount for a set of input prices
+      // and double check the payout maps to what is expected.
+
+      const fixedPointAdjustment = toBN(toWei("1"));
+
+      // Input a range of prices and check the library returns the expected value. The equation below uses the financial
+      // form of the range bond equation where as the library uses an algebraic simplification of this equation. This
+      // test validates the correct mapping between these two forms.
+      for (const price of [toWei("5.555"), toWei("11"), toWei("33"), toWei("55"), toWei("66"), toWei("111")]) {
+        const expiraryTokensForCollateral = await rangeBondCFDFPL.computeExpiraryTokensForCollateral.call(price, {
+          from: cfdMock.address,
+        });
+        //
+        const term1 = BN.min(
+          toBN(bondNotional).mul(fixedPointAdjustment).div(toBN(price)),
+          toBN(bondNotional).mul(fixedPointAdjustment).div(toBN(lowPriceRange))
+        );
+
+        const term2 = BN.max(
+          toBN(bondNotional)
+            .div(toBN(highPriceRange))
+            .mul(fixedPointAdjustment)
+            .mul(toBN(price).sub(toBN(highPriceRange)))
+            .div(toBN(price))
+            .addn(1),
+          toBN("0")
+        );
```

Comment:
This is awesome, as it verifies that the implementation matches the spec equation. +1